pkgs/by-name/ld/ld64/package.nix

```diff
@@ -0,0 +1,184 @@
+  version = "951.9";
+
+  outputs = [
+    "out"
+    "dev"
+    "lib"
+  ];
+
+  src = fetchFromGitHub {
+    owner = "apple-oss-distributions";
+    repo = "ld64";
+    rev = "ld64-${finalAttrs.version}";
+    hash = "sha256-hLkfqgBwVPlO4gfriYOawTO5E1zSD63ZcNetm1E5I70";
+  };
+
+  sourceRoot = "source";
+
+  postUnpack = ''
+    unpackFile '${xnu}'
+  '';
+
+  patches = [
+    # Use std::atomic for atomics.
+    ./0004-Use-std-atomics-and-std-mutex.patch
+    # ld64 assumes the default libLTO.dylib can be found relative to its bindir, which is
+    # not the case in nixpkgs. Override it to default to `stdenv.cc`’s libLTO.dylib.
+    ./0005-Support-LTO-in-nixpkgs.patch
+    # Add implementation of missing function required for code directory support.
+    ./0006-Add-libcd_is_blob_a_linker_signature-implementation.patch
+    # Add OpenSSL implementation of CoreCrypto digest functions
+    ./0007-Add-OpenSSL-based-CoreCrypto-digest-functions.patch
+    # ld64 will search `/usr/lib`, `/Library/Frameworks`, etc by default. Disable that.
+    ./0008-Disable-searching-in-standard-library-locations.patch
+  ];
+
+  postPatch = ''
+    substitute ${./meson.build} meson.build \
+      --subst-var-by targetPrefix '${targetPrefix}' \
+      --subst-var version
+
+    # Copy headers for certain private APIs
+    mkdir -p include
+    substitute ${crashreporter_h} include/CrashReporterClient.h \
+      --replace-fail 'USE(APPLE_INTERNAL_SDK)' '0'
+
+    # Copy from the source so the headers can be used on Linux and x86_64-darwin
+    mkdir -p include/System
+    for dir in arm i386 machine; do
+      cp -r ../xnu-src/osfmk/$dir include/System/$dir
+    done
+    mkdir -p include/sys
+    cp ../xnu-src/bsd/sys/commpage.h include/sys
+
+    # Match the version format used by upstream.
+    sed -i src/ld/Options.cpp \
+      -e '1iconst char ld_classicVersionString[] = "@(#)PROGRAM:ld PROJECT:ld64-${finalAttrs.version}\\n";'
+
+    # Instead of messing around with trying to extract and run the script from the Xcode project,
+    # just use our own Python script to generate `compile_stubs.h`
+    cp ${./gen_compile_stubs.py} gen_compile_stubs.py
+
+    # Enable LTO support using LLVM’s libLTO.dylib by default.
+    substituteInPlace src/ld/InputFiles.cpp \
+      --subst-var-by libllvm '${lib.getLib llvm}'
+    substituteInPlace src/ld/parsers/lto_file.cpp \
+      --subst-var-by libllvm '${lib.getLib llvm}'
+
+    # Use portable includes
+    substituteInPlace src/ld/code-sign-blobs/endian.h \
+      --replace-fail '#include <machine/endian.h>' '#include <sys/types.h>'
+  '';
+
+  strictDeps = true;
+
+  nativeBuildInputs = [
+    meson
+    ninja
+    pkg-config
+    python3
+  ];
+
+  buildInputs = [
+    libtapi
+    llvm
+    libunwind
+    openssl
+    xar
+  ] ++ lib.optionals stdenv.isDarwin [ darwin.dyld ] ++ lib.optionals stdenv.isLinux [ libdispatch ];
+
+  mesonBuildType = "release";
+
+  mesonFlags = [
+    (lib.mesonOption "b_ndebug" "if-release")
```

Comment:
ld64 uses `NDEBUG` in places. The upstream Xcode project defines it in the project. Specifying `b_ndebug` is how Meson sets it. Should I also add a note that enabling a debug build will cause assertion failures due to UB?

Reply:
yeah, maybe a comment in case someone tries to compile with asserts enabled. tho i haven't use meson much so if it is generally obvious feel free to leave it.

Reply:
I added a comment explaining about the UB and why debug builds don’t work.